tests/service/iam: IAM Account Alias depends_on, Import, and Data Source Testing

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Description

There's a few things with the `aws_iam_account_alias` resource and data source testing that could be improved:

```
=== RUN   TestAccAWSIAMAccountAlias_serial/Basic/basic
resource_aws_iam_account_alias_test.go:66: Step 2/2 error: Expected a non-empty plan, but got an empty plan
--- FAIL: TestAccAWSIAMAccountAlias_serial/Basic/basic (77.68s)
```

This is because the "basic" configuration includes both the resource and the data source. Typically, data source testing is done separately. Also, the data source configuration must either contain `depends_on` or be split into two separate test steps since there is no argument available to setup operation ordering otherwise. Terraform CLI version 0.14 no longer returns a non-empty plan with data source `depends_on` usage, so `ExpectNonEmptyPlan` should be removed (its fine if that testing fails on earlier versions at this point).

The import test can be folded into the basic test as is conventional with the rest of the provider.

The custom TestCheckFunc `testAccCheckAWSIAMAccountAliasDataExists` can be replaced with `TestCheckResourceAttrPair`.

### Affected Resource

* aws_iam_account_alias

--- a/aws/data_source_aws_iam_account_alias_test.go
+++ b/aws/data_source_aws_iam_account_alias_test.go
@@ -1,0 +1,44 @@
+package aws
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func testAccAWSIAMAccountAliasDataSource_basic(t *testing.T) {
+	dataSourceName := "data.aws_iam_account_alias.test"
+	resourceName := "aws_iam_account_alias.test"
+
+	rName := acctest.RandomWithPrefix("tf-acc-test")
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		ErrorCheck:   testAccErrorCheck(t, iam.EndpointsID),
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSIAMAccountAliasDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSIAMAccountAliasDataSourceConfig(rName),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttrPair(dataSourceName, "account_alias", resourceName, "account_alias"),
+				),
+			},
+		},
+	})
+}
+
+func testAccAWSIAMAccountAliasDataSourceConfig(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_iam_account_alias" "test" {
+  account_alias = %[1]q
+}
+
+data "aws_iam_account_alias" "test" {
+  depends_on = [aws_iam_account_alias.test]
+}
+`, rName)
+}

--- a/aws/resource_aws_iam_account_alias_test.go
+++ b/aws/resource_aws_iam_account_alias_test.go
@@ -13,11 +13,11 @@ import (
 
 func TestAccAWSIAMAccountAlias_serial(t *testing.T) {
 	testCases := map[string]map[string]func(t *testing.T){
-		"Basic": {
-			"basic": testAccAWSIAMAccountAlias_basic_with_datasource,
+		"DataSource": {
+			"basic": testAccAWSIAMAccountAliasDataSource_basic,
 		},
-		"Import": {
-			"import": testAccAWSIAMAccountAlias_importBasic,
+		"Resource": {
+			"basic": testAccAWSIAMAccountAlias_basic,
 		},
 	}
 
@@ -34,10 +34,10 @@ func TestAccAWSIAMAccountAlias_serial(t *testing.T) {
 	}
 }
 
-func testAccAWSIAMAccountAlias_importBasic(t *testing.T) {
+func testAccAWSIAMAccountAlias_basic(t *testing.T) {
 	resourceName := "aws_iam_account_alias.test"
 
-	rstring := acctest.RandString(5)
+	rName := acctest.RandomWithPrefix("tf-acc-test")
 
 	resource.Test(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
@@ -46,46 +46,13 @@ func testAccAWSIAMAccountAlias_importBasic(t *testing.T) {
 		CheckDestroy: testAccCheckAWSIAMAccountAliasDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAWSIAMAccountAliasConfig(rstring),
+				Config: testAccAWSIAMAccountAliasConfig(rName),
 			},
 
 			{
 				ResourceName:      resourceName,
 				ImportState:       true,
 				ImportStateVerify: true,
-			},
-		},
-	})
-}
-
-func testAccAWSIAMAccountAlias_basic_with_datasource(t *testing.T) {
-	var account_alias string
-
-	rstring := acctest.RandString(5)
-
-	resource.Test(t, resource.TestCase{
-		PreCheck:     func() { testAccPreCheck(t) },
-		ErrorCheck:   testAccErrorCheck(t, iam.EndpointsID),
-		Providers:    testAccProviders,
-		CheckDestroy: testAccCheckAWSIAMAccountAliasDestroy,
-		Steps: []resource.TestStep{
-			{
-				Config: testAccAWSIAMAccountAliasConfig(rstring),
-				Check: resource.ComposeTestCheckFunc(
-					testAccCheckAWSIAMAccountAliasExists("aws_iam_account_alias.test", &account_alias),
-				),
-			},
-			{
-				Config: testAccAWSIAMAccountAliasConfig_with_datasource(rstring),
-				Check: resource.ComposeTestCheckFunc(
-					testAccCheckAWSIAMAccountAliasExists("aws_iam_account_alias.test", &account_alias),
-					testAccCheckAWSIAMAccountAliasDataExists("data.aws_iam_account_alias.current", &account_alias),
-				),
-				// We expect a non-empty plan due to the way data sources and depends_on
-				// work, or don't work. See https://github.com/hashicorp/terraform/issues/11139#issuecomment-275121893
-				// We accept this limitation and feel this test is OK because of the
-				// explicitly check above
-				ExpectNonEmptyPlan: true,
 			},
 		},
 	})
@@ -116,21 +83,6 @@ func testAccCheckAWSIAMAccountAliasDestroy(s *terraform.State) error {
 
 }
 
-func testAccCheckAWSIAMAccountAliasDataExists(n string, a *string) resource.TestCheckFunc {
-	return func(s *terraform.State) error {
-		rs, ok := s.RootModule().Resources[n]
-		if !ok {
-			return fmt.Errorf("Not found: %s", n)
-		}
-
-		if rs.Primary.Attributes["account_alias"] != *a {
-			return fmt.Errorf("Data Source account_alias didn't match, expected (%s), got (%s)", *a, rs.Primary.Attributes["account_alias"])
-		}
-
-		return nil
-	}
-}
-
 func testAccCheckAWSIAMAccountAliasExists(n string, a *string) resource.TestCheckFunc {
 	return func(s *terraform.State) error {
 		rs, ok := s.RootModule().Resources[n]
@@ -157,22 +109,10 @@ func testAccCheckAWSIAMAccountAliasExists(n string, a *string) resource.TestChec
 	}
 }
 
-func testAccAWSIAMAccountAliasConfig_with_datasource(rstring string) string {
+func testAccAWSIAMAccountAliasConfig(rName string) string {
 	return fmt.Sprintf(`
 resource "aws_iam_account_alias" "test" {
-  account_alias = "terraform-%s-alias"
+  account_alias = %[1]q
 }
-
-data "aws_iam_account_alias" "current" {
-  depends_on = [aws_iam_account_alias.test]
-}
-`, rstring)
-}
-
-func testAccAWSIAMAccountAliasConfig(rstring string) string {
-	return fmt.Sprintf(`
-resource "aws_iam_account_alias" "test" {
-  account_alias = "terraform-%s-alias"
-}
-`, rstring)
+`, rName)
 }

--- a/aws/resource_aws_iam_account_alias_test.go
+++ b/aws/resource_aws_iam_account_alias_test.go
@@ -4,7 +4,6 @@ import (
 	"fmt"
 	"testing"
 
-	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/iam"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
@@ -47,8 +46,10 @@ func testAccAWSIAMAccountAlias_basic(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: testAccAWSIAMAccountAliasConfig(rName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckAWSIAMAccountAliasExists(resourceName),
+				),
 			},
-
 			{
 				ResourceName:      resourceName,
 				ImportState:       true,
@@ -83,7 +84,7 @@ func testAccCheckAWSIAMAccountAliasDestroy(s *terraform.State) error {
 
 }
 
-func testAccCheckAWSIAMAccountAliasExists(n string, a *string) resource.TestCheckFunc {
+func testAccCheckAWSIAMAccountAliasExists(n string) resource.TestCheckFunc {
 	return func(s *terraform.State) error {
 		rs, ok := s.RootModule().Resources[n]
 		if !ok {
@@ -102,8 +103,6 @@ func testAccCheckAWSIAMAccountAliasExists(n string, a *string) resource.TestChec
 		if len(resp.AccountAliases) == 0 {
 			return fmt.Errorf("Bad: Account alias %q does not exist", rs.Primary.ID)
 		}
-
-		*a = aws.StringValue(resp.AccountAliases[0])
 
 		return nil
 	}

